pyaedt/edb_core/edb_data/sources.py

```diff
@@ -506,14 +534,308 @@ def pec_launch_width(self):
+        """
+
+        refTerm = self._edb_terminal.GetReferenceTerminal()
+        if self._edb_terminal.GetTerminalType() == self._pedb.edb.Cell.Terminal.TerminalType.PinGroupTerminal:
+            padStackInstance = self._edb_terminal.GetPinGroup().GetPins()[0]
+            pingroup = refTerm.GetPinGroup()
+            refPinList = pingroup.GetPins()
+            return self._get_closest_pin(padStackInstance, refPinList, gnd_net_name_preference)
+        elif self._edb_terminal.GetTerminalType() == self._pedb.edb.Cell.Terminal.TerminalType.PadstackInstanceTerminal:
+            _, padStackInstance, layer = self._edb_terminal.GetParameters()
+            if refTerm.GetTerminalType() == self._pedb.edb.Cell.Terminal.TerminalType.PinGroupTerminal:
+                pingroup = refTerm.GetPinGroup()
+                refPinList = pingroup.GetPins()
+                return self._get_closest_pin(padStackInstance, refPinList, gnd_net_name_preference)
+            else:
+                try:
+                    returnOK, refTermPSI, layer = refTerm.GetParameters()
+                    return EDBPadstackInstance(refTermPSI, self._pedb)
+                except AttributeError:
+                    return None
+        return None  # pragma: no cover
+
+    @pyaedt_function_handler()
+    def get_edge_terminal_reference_primitive(self):
+        """This returns a string primitive instance instances and serves Edge ports,
+         wave ports and coupled edge ports that are directly connedted to primitives
+
+        Returns
+        -------
+        :class:`pyaedt.edb_core.edb_data.primitives_data.EDBPrimitives`
+        """
+
+        ref_layer = self._edb_terminal.GetReferenceLayer()
+        edges = self._edb_terminal.GetEdges()
+        _, prim_value, point_data = edges[0].GetParameters()
+        X = point_data.X
+        Y = point_data.Y
+        shape_pd = self._pedb.edb.Geometry.PointData(X, Y)
+        layer_name = ref_layer.GetName()
+        for primitive in self._pedb.active_layout.Primitives:
+            if primitive.GetLayer().GetName() == layer_name or not layer_name:
+                prim_shape_data = primitive.GetPolygonData()
+                if prim_shape_data.PointInPolygon(shape_pd):
+                    return EDBPrimitives(primitive, self._pedb)
+        return None  # pragma: no cover
+
+    @pyaedt_function_handler()
+    def get_point_terminal_reference_primitive(self):
+        """This returns the primitive reference for the point terminal.
+
+        Returns
+        -------
+        :class:`pyaedt.edb_core.edb_data.padstacks_data.EDBPadstackInstance` or
+        :class:`pyaedt.edb_core.edb_data.primitives_data.EDBPrimitives`
+        """
+
+        ref_term = self._edb_terminal.GetReferenceTerminal()  # return value is type terminal
+        _, point_data, layer = ref_term.GetParameters()
+        X = point_data.X
+        Y = point_data.Y
+        shape_pd = self._pedb.edb.Geometry.PointData(X, Y)
```

Comment:
No need to create a variable if it is only used once and can be easily evaluated.
```suggestion
        shape_pd = self._pedb.edb.Geometry.PointData(point_data.X, point_data.Y)
```